libraries/concurrent-merkle-tree/src/cmt.rs

```diff
@@ -0,0 +1,447 @@
+{
+}
+unsafe impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> Pod
+    for ConcurrentMerkleTree<MAX_DEPTH, MAX_BUFFER_SIZE>
+{
+}
+
+impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> Default
+    for ConcurrentMerkleTree<MAX_DEPTH, MAX_BUFFER_SIZE>
+{
+    fn default() -> Self {
+        Self {
+            sequence_number: 0,
+            active_index: 0,
+            buffer_size: 0,
+            change_logs: [ChangeLog::<MAX_DEPTH>::default(); MAX_BUFFER_SIZE],
+            rightmost_proof: Path::<MAX_DEPTH>::default(),
+        }
+    }
+}
+
+impl<const MAX_DEPTH: usize, const MAX_BUFFER_SIZE: usize> ConcurrentMerkleTree<MAX_DEPTH, MAX_BUFFER_SIZE> {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// This is the trustless initialization method that should be used in most cases.
+    pub fn initialize(&mut self) -> Result<Node, CMTError> {
+        check_bounds(MAX_DEPTH, MAX_BUFFER_SIZE);
+        let mut rightmost_proof = Path::default();
+        let mut empty_node_cache = Box::new([Node::default(); MAX_DEPTH]);
+        for (i, node) in rightmost_proof.proof.iter_mut().enumerate() {
+            *node = empty_node_cached::<MAX_DEPTH>(i as u32, &mut empty_node_cache);
+        }
+        let mut path = [Node::default(); MAX_DEPTH];
+        for (i, node) in path.iter_mut().enumerate() {
+            *node = empty_node_cached::<MAX_DEPTH>(i as u32, &mut empty_node_cache);
+        }
+        self.change_logs[0].root = empty_node(MAX_DEPTH as u32);
+        self.change_logs[0].path = path;
+        self.sequence_number = 0;
+        self.active_index = 0;
+        self.buffer_size = 1;
+        self.rightmost_proof = rightmost_proof;
+        Ok(self.change_logs[0].root)
+    }
+
+    /// This is a trustful initialization method that assumes the root contains the expected
+    /// leaves.
+    ///
+    /// At the time of this crate's publishing, there is no supported way to efficiently verify
+    /// a pre-initialized root on-chain. Using this method before having a method for on-chain verification
+    /// will prevent other applications from indexing the leaf data stored in this tree.
+    pub fn initialize_with_root(
+        &mut self,
+        root: Node,
+        rightmost_leaf: Node,
+        proof_vec: &[Node],
+        index: u32,
+    ) -> Result<Node, CMTError> {
+        check_bounds(MAX_DEPTH, MAX_BUFFER_SIZE);
+        let mut proof: [Node; MAX_DEPTH] = [Node::default(); MAX_DEPTH];
+        proof.copy_from_slice(proof_vec);
+        let rightmost_proof = Path {
+            proof,
+            index: index + 1,
+            leaf: rightmost_leaf,
+            _padding: 0,
+        };
+        self.change_logs[0].root = root;
+        self.sequence_number = 1;
+        self.active_index = 0;
+        self.buffer_size = 1;
+        self.rightmost_proof = rightmost_proof;
+        assert_eq!(root, recompute(rightmost_leaf, &proof, index,));
```

Comment:
Yes this can fail. This is complex to reason about in the practical case, so I expect this function to change quite a bit in the coming months.

This now returns `InvalidProof`